core/builtin.py

```diff
@@ -526,7 +531,13 @@ def Cd(argv, mem, dir_stack):
     util.error("cd %r: %s", dest_dir, os.strerror(e.errno))
     return 1
 
-  state.SetGlobalString(mem, 'PWD', dest_dir)  # Set $PWD.
+  # Set $PWD.
+  if arg.P:
+    state.SetGlobalString(mem, 'PWD', os.path.realpath(dest_dir))
+  else: # `-L` is the default behavior; no need to check it
+    # TODO: ensure that if multiple flags are provided, the *last* one overrides
+    # the others
+    state.SetGlobalString(mem, 'PWD', dest_dir)
```

Comment:
I would factor out the dir like:

    pwd = os.path.realpath(dest_dir) if arg.P else dest_dir
    state.Set...